File: geo/harvesters/datagouv.py
from __future__ import annotations

import csv
import io
import json
import logging
from collections.abc import Iterable
from typing import Any

import httpx

from config import get_settings
from harvesters.common import build_address, first_string, parse_datetime
from models import Event
from services.debug import persist_debug_artifact
from services.ban import resolve_address

logger = logging.getLogger(__name__)

DATASET_KEYWORDS = (
    "agenda",
    "evenement",
    "événement",
    "manifestation",
    "culture",
    "festival",
    "mairie",
)

SEARCH_QUERIES = (
    "agenda",
    "evenement",
    "culture",
    "festival",
    "mairie",
)


def _matches_keywords(*values: str) -> bool:
    haystack = " ".join(values).lower()
    return any(keyword in haystack for keyword in DATASET_KEYWORDS)


def _pick_field(row: dict[str, object], *names: str) -> str:
    lowered = {key.lower(): key for key in row}
    for candidate in names:
        if candidate in lowered:
            return first_string(row[lowered[candidate]])
    return ""


def _iter_rows_from_json(payload: object) -> Iterable[dict[str, object]]:
    if isinstance(payload, list):
        for item in payload:
            if isinstance(item, dict):
                yield item
        return
    if isinstance(payload, dict):
        for key in ("data", "results", "records", "items"):
            value = payload.get(key)
            if isinstance(value, list):
                for item in value:
                    if isinstance(item, dict):
                        yield item


def _record_row_failure(
    *,
    row: dict[str, object],
    source_uid: str,
    source_url: str,
    exc: Exception,
) -> str:
    return persist_debug_artifact(
        source_tag="datagouv",
        artifact_type="datagouv_raw",
        stage="normalize",
        identifier=source_uid,
        payload=row,
        metadata={
            "error": str(exc),
            "source_uid": source_uid,
            "source_url": source_url,
        },
    )


async def normalize_datagouv_row(
    row: dict[str, object],
    source_uid: str,
    source_url: str,
) -> Event | None:
    title = _pick_field(row, "title", "titre", "name", "nom", "event", "summary")
    start_dt = parse_datetime(
        _pick_field(
            row,
            "start_dt",
            "start_date",
            "date_start",
            "date_debut",
            "date",
            "datetime",
        )
    )
    address = build_address(
        [
            _pick_field(row, "address", "adresse", "lieu", "location"),
            _pick_field(row, "postal_code", "code_postal", "postcode"),
            _pick_field(row, "city", "ville", "commune"),
        ]
    )

    if not title or start_dt is None or not address:
        return None

    lon_text = _pick_field(row, "longitude", "lon", "lng", "x")
    lat_text = _pick_field(row, "latitude", "lat", "y")
    if lon_text and lat_text:
        coordinates = (float(lon_text.replace(",", ".")), float(lat_text.replace(",", ".")))
    else:
        coordinates = await resolve_address(address)

    return Event(
        source_uid=source_uid,
        title=title,
        description=_pick_field(row, "description", "details", "resume", "summary"),
        start_dt=start_dt,
        end_dt=parse_datetime(
            _pick_field(row, "end_dt", "end_date", "date_fin", "end")
        ),
        location_name=_pick_field(row, "location_name", "venue", "place", "lieu"),
        address=address,
        longitude=coordinates[0],
        latitude=coordinates[1],
        source_tag="datagouv",
        source_url=source_url,
    )


async def _parse_resource(client: httpx.AsyncClient, resource: dict) -> list[Event]:
    url = resource.get("url")
    if not isinstance(url, str) or not url:
        return []

    response = await client.get(url)
    response.raise_for_status()

    format_name = first_string(resource.get("format")).lower()
    source_uid_prefix = str(resource.get("id") or url)
    events: list[Event] = []

    if format_name == "csv":
        content = response.text
        reader = csv.DictReader(io.StringIO(content))
        for index, row in enumerate(reader):
            source_uid = f"{source_uid_prefix}:{index}"
            try:
                event = await normalize_datagouv_row(row, source_uid, url)
            except Exception as exc:
                artifact_id = _record_row_failure(
                    row=row,
                    source_uid=source_uid,
                    source_url=url,
                    exc=exc,
                )
                logger.warning(
                    "Skipping data.gouv record %s: %s [artifact_id=%s]",
                    source_uid,
                    exc,
                    artifact_id,
                )
                continue
            if event is not None:
                events.append(event)
        return events

    if format_name in {"json", "geojson"}:
        payload = response.json()
        for index, row in enumerate(_iter_rows_from_json(payload)):
            source_uid = f"{source_uid_prefix}:{index}"
            try:
                event = await normalize_datagouv_row(row, source_uid, url)
            except Exception as exc:
                artifact_id = _record_row_failure(
                    row=row,
                    source_uid=source_uid,
                    source_url=url,
                    exc=exc,
                )
                logger.warning(
                    "Skipping data.gouv record %s: %s [artifact_id=%s]",
                    source_uid,
                    exc,
                    artifact_id,
                )
                continue
            if event is not None:
                events.append(event)
        return events

    content_type = response.headers.get("content-type", "")
    if "json" in content_type:
        payload = json.loads(response.text)
        for index, row in enumerate(_iter_rows_from_json(payload)):
            source_uid = f"{source_uid_prefix}:{index}"
            try:
                event = await normalize_datagouv_row(row, source_uid, url)
            except Exception as exc:
                artifact_id = _record_row_failure(
                    row=row,
                    source_uid=source_uid,
                    source_url=url,
                    exc=exc,
                )
                logger.warning(
                    "Skipping data.gouv record %s: %s [artifact_id=%s]",
                    source_uid,
                    exc,
                    artifact_id,
                )
                continue
            if event is not None:
                events.append(event)

    return events


async def replay_datagouv_payload(
    payload: dict[str, Any],
    *,
    source_uid: str,
    source_url: str,
) -> Event | None:
    if not isinstance(payload, dict):
        raise ValueError("data.gouv replay payload must be an object")
    return await normalize_datagouv_row(payload, source_uid, source_url)


async def fetch_datagouv_events() -> list[Event]:
    settings = get_settings()
    events: list[Event] = []
    seen_dataset_ids: set[str] = set()
    seen_resource_ids: set[str] = set()

    async with httpx.AsyncClient(
        base_url=settings.datagouv_api_url,
        timeout=settings.http_timeout_seconds,
        follow_redirects=True,
    ) as client:
        total_datasets = 0
        for query in SEARCH_QUERIES:
            response = await client.get("/datasets/", params={"page_size": 20, "q": query})
            response.raise_for_status()
            payload = response.json()
            datasets = payload.get("data") or payload.get("results") or []
            total_datasets += len(datasets)

            for dataset in datasets:
                if not isinstance(dataset, dict):
                    continue
                dataset_id = str(dataset.get("id") or dataset.get("slug") or "")
                if dataset_id and dataset_id in seen_dataset_ids:
                    continue
                if dataset_id:
                    seen_dataset_ids.add(dataset_id)

                title = first_string(dataset.get("title"))
                description = first_string(dataset.get("description"))
                if not _matches_keywords(title, description):
                    continue

                resources = dataset.get("resources") or []
                for resource in resources:
                    if not isinstance(resource, dict):
                        continue
                    resource_id = str(resource.get("id") or resource.get("url") or "")
                    if resource_id and resource_id in seen_resource_ids:
                        continue
                    if resource_id:
                        seen_resource_ids.add(resource_id)

                    format_name = first_string(resource.get("format")).lower()
                    if format_name not in {"csv", "json", "geojson"}:
                        continue
                    try:
                        events.extend(await _parse_resource(client, resource))
                    except Exception as exc:
                        logger.warning(
                            "Skipping data.gouv resource %s: %s",
                            resource.get("id") or resource.get("url"),
                            exc,
                        )

        if total_datasets == 0:
            logger.warning(
                "data.gouv search returned no datasets for queries=%s",
                ",".join(SEARCH_QUERIES),
            )

    return events

File: back/internal/ingestion/sources/datagouv/datagouv.go
package datagouv

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"strconv"
	"strings"

	"github.com/tanguyRa/onvaou/internal/config"
	"github.com/tanguyRa/onvaou/internal/ingestion/artifacts"
	"github.com/tanguyRa/onvaou/internal/ingestion/geocoder"
	"github.com/tanguyRa/onvaou/internal/ingestion/model"
	"github.com/tanguyRa/onvaou/internal/ingestion/store"
	"github.com/tanguyRa/onvaou/internal/ingestion/util"
)

var (
	datasetKeywords = []string{"agenda", "evenement", "événement", "manifestation", "culture", "festival", "mairie"}
)

type Service struct {
	cfg       config.Config
	logger    *slog.Logger
	client    *http.Client
	geocoder  *geocoder.BANClient
	store     *store.Store
	artifacts *artifacts.Store
}

func NewService(cfg config.Config, logger *slog.Logger, client *http.Client, geocoder *geocoder.BANClient, store *store.Store, artifactStore *artifacts.Store) *Service {
	return &Service{
		cfg:       cfg,
		logger:    logger,
		client:    client,
		geocoder:  geocoder,
		store:     store,
		artifacts: artifactStore,
	}
}

func (s *Service) Run(ctx context.Context) (model.IngestionResult, error) {
	events, err := s.fetchEvents(ctx)
	if err != nil {
		return model.IngestionResult{}, err
	}
	return s.store.UpsertEventBatch(ctx, "datagouv", events)
}

func (s *Service) ReplayPayload(ctx context.Context, payload interface{}, metadata map[string]interface{}) (*model.Event, error) {
	row, ok := payload.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("data.gouv replay payload must be an object")
	}
	sourceUID := fmt.Sprint(metadata["source_uid"])
	sourceURL := fmt.Sprint(metadata["source_url"])
	return s.normalizeRow(ctx, row, sourceUID, sourceURL)
}

func (s *Service) fetchEvents(ctx context.Context) ([]model.Event, error) {
	endpoint, err := url.Parse(s.cfg.Ingestion.DataGouvAPIURL)
	if err != nil {
		return nil, err
	}

	queries := s.cfg.Ingestion.DataGouvQueries
	if len(queries) == 0 {
		queries = []string{"agenda", "evenement", "culture", "festival", "mairie"}
	}

	events := make([]model.Event, 0)
	seenDatasetIDs := map[string]struct{}{}
	seenResourceIDs := map[string]struct{}{}
	totalDatasets := 0

	for _, query := range queries {
		searchURL := *endpoint
		searchURL.Path += "/datasets/"
		params := searchURL.Query()
		params.Set("page_size", "20")
		params.Set("q", query)
		searchURL.RawQuery = params.Encode()

		req, err := http.NewRequestWithContext(ctx, http.MethodGet, searchURL.String(), nil)
		if err != nil {
			return nil, err
		}

		resp, err := s.client.Do(req)
		if err != nil {
			return nil, err
		}
		if resp.StatusCode >= 400 {
			resp.Body.Close()
			return nil, fmt.Errorf("data.gouv upstream returned status %d", resp.StatusCode)
		}

		var payload map[string]interface{}
		err = json.NewDecoder(resp.Body).Decode(&payload)
		resp.Body.Close()
		if err != nil {
			return nil, err
		}

		rawDatasets := extractRows(payload)
		totalDatasets += len(rawDatasets)

		for _, dataset := range rawDatasets {
			datasetID := fallbackString(fmt.Sprint(dataset["id"]), fmt.Sprint(dataset["slug"]))
			if datasetID != "" {
				if _, exists := seenDatasetIDs[datasetID]; exists {
					continue
				}
				seenDatasetIDs[datasetID] = struct{}{}
			}

			title := util.FirstString(dataset["title"])
			description := util.FirstString(dataset["description"])
			if !matchesKeywords(title, description) {
				continue
			}

			resources, _ := dataset["resources"].([]interface{})
			for _, rawResource := range resources {
				resource, ok := rawResource.(map[string]interface{})
				if !ok {
					continue
				}
				resourceID := fallbackString(fmt.Sprint(resource["id"]), fmt.Sprint(resource["url"]))
				if resourceID != "" {
					if _, exists := seenResourceIDs[resourceID]; exists {
						continue
					}
					seenResourceIDs[resourceID] = struct{}{}
				}

				formatName := strings.ToLower(util.FirstString(resource["format"]))
				if formatName != "csv" && formatName != "json" && formatName != "geojson" {
					continue
				}

				parsedEvents, err := s.parseResource(ctx, resource)
				if err != nil {
					s.logger.Warn("skipping data.gouv resource", "resource", resourceID, "error", err)
					continue
				}
				events = append(events, parsedEvents...)
			}
		}
	}

	if totalDatasets == 0 {
		s.logger.Warn("data.gouv search returned no datasets", "queries", strings.Join(queries, ","))
	}

	return events, nil
}

func (s *Service) parseResource(ctx context.Context, resource map[string]interface{}) ([]model.Event, error) {
	rawURL, ok := resource["url"].(string)
	if !ok || rawURL == "" {
		return nil, nil
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, err
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("resource upstream returned status %d", resp.StatusCode)
	}

	formatName := strings.ToLower(util.FirstString(resource["format"]))
	sourceUIDPrefix := fallbackString(fmt.Sprint(resource["id"]), rawURL)

	switch formatName {
	case "csv":
		return s.parseCSVResource(ctx, resp.Body, sourceUIDPrefix, rawURL)
	case "json", "geojson":
		return s.parseJSONResource(ctx, resp.Body, sourceUIDPrefix, rawURL)
	default:
		contentType := resp.Header.Get("Content-Type")
		if strings.Contains(strings.ToLower(contentType), "json") {
			return s.parseJSONResource(ctx, resp.Body, sourceUIDPrefix, rawURL)
		}
	}

	return nil, nil
}

func (s *Service) parseCSVResource(ctx context.Context, body io.Reader, sourceUIDPrefix string, sourceURL string) ([]model.Event, error) {
	reader := csv.NewReader(body)
	rows, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return nil, nil
	}

	headers := rows[0]
	events := make([]model.Event, 0, len(rows)-1)
	for index, rowValues := range rows[1:] {
		row := map[string]interface{}{}
		for i, header := range headers {
			if i < len(rowValues) {
				row[header] = rowValues[i]
			}
		}
		sourceUID := fmt.Sprintf("%s:%d", sourceUIDPrefix, index)
		event, err := s.normalizeRow(ctx, row, sourceUID, sourceURL)
		if err != nil {
			artifactID, artifactErr := s.artifacts.Persist(
				"datagouv",
				"datagouv_raw",
				"normalize",
				sourceUID,
				row,
				map[string]interface{}{
					"error":      err.Error(),
					"source_uid": sourceUID,
					"source_url": sourceURL,
				},
			)
			if artifactErr == nil {
				s.logger.Warn("skipping data.gouv record", "source_uid", sourceUID, "error", err, "artifact_id", artifactID)
			} else {
				s.logger.Warn("skipping data.gouv record", "source_uid", sourceUID, "error", err)
			}
			continue
		}
		if event != nil {
			events = append(events, *event)
		}
	}
	return events, nil
}

func (s *Service) parseJSONResource(ctx context.Context, body io.Reader, sourceUIDPrefix string, sourceURL string) ([]model.Event, error) {
	var payload interface{}
	if err := json.NewDecoder(body).Decode(&payload); err != nil {
		return nil, err
	}

	rows := extractRows(payload)
	events := make([]model.Event, 0, len(rows))
	for index, row := range rows {
		sourceUID := fmt.Sprintf("%s:%d", sourceUIDPrefix, index)
		event, err := s.normalizeRow(ctx, row, sourceUID, sourceURL)
		if err != nil {
			artifactID, artifactErr := s.artifacts.Persist(
				"datagouv",
				"datagouv_raw",
				"normalize",
				sourceUID,
				row,
				map[string]interface{}{
					"error":      err.Error(),
					"source_uid": sourceUID,
					"source_url": sourceURL,
				},
			)
			if artifactErr == nil {
				s.logger.Warn("skipping data.gouv record", "source_uid", sourceUID, "error", err, "artifact_id", artifactID)
			} else {
				s.logger.Warn("skipping data.gouv record", "source_uid", sourceUID, "error", err)
			}
			continue
		}
		if event != nil {
			events = append(events, *event)
		}
	}
	return events, nil
}

func (s *Service) normalizeRow(ctx context.Context, row map[string]interface{}, sourceUID string, sourceURL string) (*model.Event, error) {
	title := pickField(row, "title", "titre", "name", "nom", "event", "summary")
	startDT := util.ParseDateTime(
		pickField(row, "start_dt", "start_date", "date_start", "date_debut", "date", "datetime"),
	)
	address := util.BuildAddress(
		pickField(row, "address", "adresse", "lieu", "location"),
		pickField(row, "postal_code", "code_postal", "postcode"),
		pickField(row, "city", "ville", "commune"),
	)
	if title == "" || startDT == nil || address == "" {
		return nil, nil
	}

	lonText := pickField(row, "longitude", "lon", "lng", "x")
	latText := pickField(row, "latitude", "lat", "y")

	var lon float64
	var lat float64
	if lonText != "" && latText != "" {
		var err error
		lon, err = strconv.ParseFloat(strings.ReplaceAll(lonText, ",", "."), 64)
		if err != nil {
			return nil, err
		}
		lat, err = strconv.ParseFloat(strings.ReplaceAll(latText, ",", "."), 64)
		if err != nil {
			return nil, err
		}
	} else {
		var err error
		lon, lat, err = s.geocoder.ResolveAddress(ctx, address)
		if err != nil {
			return nil, err
		}
	}

	return &model.Event{
		SourceUID:    util.NormalizeSpace(sourceUID),
		Title:        title,
		Description:  pickField(row, "description", "details", "resume", "summary"),
		StartDT:      *startDT,
		EndDT:        util.ParseDateTime(pickField(row, "end_dt", "end_date", "date_fin", "end")),
		LocationName: pickField(row, "location_name", "venue", "place", "lieu"),
		Address:      address,
		Longitude:    lon,
		Latitude:     lat,
		SourceTag:    "datagouv",
		SourceURL:    sourceURL,
	}, nil
}

func extractRows(payload interface{}) []map[string]interface{} {
	switch typed := payload.(type) {
	case []interface{}:
		out := make([]map[string]interface{}, 0, len(typed))
		for _, item := range typed {
			if row, ok := item.(map[string]interface{}); ok {
				out = append(out, row)
			}
		}
		return out
	case map[string]interface{}:
		for _, key := range []string{"data", "results", "records", "items"} {
			if rows, ok := typed[key].([]interface{}); ok {
				out := make([]map[string]interface{}, 0, len(rows))
				for _, item := range rows {
					if row, ok := item.(map[string]interface{}); ok {
						out = append(out, row)
					}
				}
				return out
			}
		}
		return []map[string]interface{}{typed}
	default:
		return nil
	}
}

func matchesKeywords(values ...string) bool {
	haystack := strings.ToLower(strings.Join(values, " "))
	for _, keyword := range datasetKeywords {
		if strings.Contains(haystack, keyword) {
			return true
		}
	}
	return false
}

func pickField(row map[string]interface{}, names ...string) string {
	lowered := make(map[string]string, len(row))
	for key := range row {
		lowered[strings.ToLower(key)] = key
	}
	for _, name := range names {
		if key, ok := lowered[name]; ok {
			return util.FirstString(row[key])
		}
	}
	return ""
}

func fallbackString(value string, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
